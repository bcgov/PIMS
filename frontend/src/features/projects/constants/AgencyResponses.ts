export enum AgencyResponses {
  // Agency does not want to receive notifications.
  Unsubscribe = 'Unsubscribe',
  // Agency wants to receive notifications.
  Subscribe = 'Subscribe',
  // Agency does not want to receive notifications, but is interested in watching the project.
  Watch = 'Watch',
}
