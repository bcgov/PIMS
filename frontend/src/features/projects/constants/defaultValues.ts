import { getCurrentFiscalYear } from 'utils/utils';

import { IProject } from '../interfaces';

export const defaultProject = (): IProject => ({
  projectNumber: '',
  name: '',
  note: '',
  notes: [],
  description: '',
  properties: [],
  tierLevelId: 1,
  statusId: 1,
  agencyId: 0,
  tasks: [],
  projectAgencyResponses: [],
  publicNote: '',
  privateNote: '',
  statusCode: 'DR',
  fiscalYear: getCurrentFiscalYear(),
  confirmation: false,
  approvedOn: '',
  deniedOn: '',
  cancelledOn: '',
  submittedOn: '',
  exemptionRequested: false,
  exemptionRationale: '',
  exemptionApprovedOn: '',
  initialNotificationSentOn: '',
  thirtyDayNotificationSentOn: '',
  sixtyDayNoficationSentOn: '',
  ninetyDayNotificationSentOn: '',
  onHoldNotificationSentOn: '',
  interestedReceivedOn: '',
  interestFromEnhancedReferralNote: '',
  transferredWithinGreOn: '',
  clearanceNotificationSentOn: '',
  requestForSplReceivedOn: '',
  approvedForSplOn: '',
  marketedOn: '',
  disposedOn: '',
  offerAcceptedOn: '',
  assessedOn: '',
  adjustedOn: '',
  offersNote: '',
  preliminaryFormSignedOn: '',
  finalFormSignedOn: '',
  netBook: '',
  assessed: '',
  appraised: '',
  market: '',
  workflowCode: '',
  offerAmount: '',
  purchaser: '',
  manager: '',
  actualFiscalYear: '',
  plannedFutureUse: '',
  preliminaryFormSignedBy: '',
  finalFormSignedBy: '',
  interestComponent: '',
  ocgFinancialStatement: '',
  salesCost: '',
  gainBeforeSpl: '',
  netProceeds: '',
  programCost: '',
  priorYearAdjustmentAmount: '',
  removalFromSplRequestOn: '',
  removalFromSplApprovedOn: '',
  removalFromSplRationale: '',
  documentationNote: '',
  salesHistoryNote: '',
  comments: '',
  loanTermsNote: '',
  closeOutNote: '',
  remediationNote: '',
  programCostNote: '',
  gainNote: '',
  adjustmentNote: '',
  statusHistory: [],
});
