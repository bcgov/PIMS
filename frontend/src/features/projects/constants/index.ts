export * from './AgencyResponses';
export * from './DisposalWorkflows';
export * from './DisposeWorkflowStatus';
export * from './ProjectWorkflowComponents';
export * from './ReviewWorkflowStatus';
export * from './SPPApprovalTabs';
