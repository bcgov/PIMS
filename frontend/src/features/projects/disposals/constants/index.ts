export * from './defaultFilter';
export * from './defaultProjectForm';
