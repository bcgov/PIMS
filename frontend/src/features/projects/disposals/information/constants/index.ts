export * from './AddPropertyColumns';
export * from './PropertyColumns';
