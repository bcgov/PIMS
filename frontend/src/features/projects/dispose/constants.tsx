import { DisposeWorkflowStatus, ProjectWorkflowComponent } from '.';
import React from 'react';

//TODO: map to real components
export const projectWorkflowComponents: ProjectWorkflowComponent[] = [
  { component: () => <p>Step 1</p>, workflowStatus: DisposeWorkflowStatus.Draft },
  { component: () => <p>Step 2</p>, workflowStatus: DisposeWorkflowStatus.SelectProperties },
  { component: () => <p>Step 3</p>, workflowStatus: DisposeWorkflowStatus.UpdateInformation },
  { component: () => <p>Step 4</p>, workflowStatus: DisposeWorkflowStatus.RequiredDocumentation },
  { component: () => <p>Step 5</p>, workflowStatus: DisposeWorkflowStatus.Approval },
  { component: () => <p>Step 6</p>, workflowStatus: DisposeWorkflowStatus.Submitted },
];
