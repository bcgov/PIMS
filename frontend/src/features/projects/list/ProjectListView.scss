$header-height: 72px;
$footer-height: 46px;
$navbar-height: 45px;
$filter-height: 76px;
$contentheading-height: 60px;

.ProjectListView {
  background-color: var(--light);
  display: flex;
  flex-direction: column;
  padding: 0;

  .filter-container {
    padding: 0;
    background-color: var(--light);
    .search-bar {
      justify-content: center;
      .bar-item {
        .form-group {
          padding: 0;
          margin: 0;
        }
      }

      .btn-warning {
        color: white;
        font-weight: bold;
      }

      .form-control {
        border-radius: 0;
      }
    }
  }

  .TableToolbar {
    display: flex;
    justify-content: flex-end;
    margin-left: auto;
    padding-bottom: 1rem;
  }

  .ScrollContainer {
    padding: 1rem 2rem;
    max-height: calc(
      100vh - #{$header-height} - #{$navbar-height} - #{$filter-height} - #{$footer-height}
    );
    overflow-y: scroll;

    table {
      background-color: #fff;
    }
  }
}
