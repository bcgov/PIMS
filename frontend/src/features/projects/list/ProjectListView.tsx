import './ProjectListView.scss';

import React, { useState, useMemo, useRef, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Container } from 'react-bootstrap';
import * as API from 'constants/API';
import { RootState } from 'reducers/rootReducer';
import { IProjectFilter, IProject } from '.';
import { columns as cols } from './columns';
import { Table } from 'components/Table';
import service from '../apiService';
import { FaFolder, FaFolderOpen, FaFileExcel, FaFileAlt } from 'react-icons/fa';
import { Properties } from './properties';
import FilterBar from 'components/SearchBar/FilterBar';
import { Col } from 'react-bootstrap';
import { Input, Button, Select } from 'components/common/form';
import GenericModal from 'components/common/GenericModal';
import { useHistory } from 'react-router-dom';
import { ReviewWorkflowStatus, IStatus, fetchProjectStatuses } from '../common';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import Claims from 'constants/claims';
import { ENVIRONMENT } from 'constants/environment';
import queryString from 'query-string';
import download from 'utils/download';
import { mapLookupCodeWithParentString, mapStatuses } from 'utils';
import styled from 'styled-components';
import TooltipWrapper from 'components/common/TooltipWrapper';
import { ParentSelect } from 'components/common/form/ParentSelect';
import variables from '_variables.module.scss';
import useCodeLookups from 'hooks/useLookupCodes';
import useDeepCompareEffect from 'hooks/useDeepCompareEffect';

interface IProjectFilterState {
  name?: string;
  statusId?: string;
  agencyId?: string;
  assessWorkflow?: boolean;
  agencies?: number;
  fiscalYaer?: number;
}

const initialValues = {
  name: '',
  statusId: '',
  agencyId: '',
};

const getProjectReportUrl = (filter: IProjectFilter) =>
  `${ENVIRONMENT.apiUrl}/reports/projects?${filter ? queryString.stringify(filter) : ''}`;

export const getProjectFinancialReportUrl = (filter?: IProjectFilter) =>
  `${ENVIRONMENT.apiUrl}/reports/projects/surplus/properties/list?${
    filter ? queryString.stringify(filter) : ''
  }`;

const FileIcon = styled(Button)`
  background-color: white !important;
  color: ${variables.primaryColor} !important;
`;

const initialQuery: IProjectFilter = {
  page: 1,
  quantity: 10,
};

export const getServerQuery = (state: {
  pageIndex: number;
  pageSize: number;
  filter: IProjectFilterState;
  agencyIds?: number[];
}) => {
  const { pageIndex, pageSize, filter: search } = state;

  const query: any = {
    ...initialQuery,
    // TODO: this field is not yet implemented in FilterBar
    page: pageIndex + 1,
    quantity: pageSize,
    ...search,
  };
  return query;
};

export enum PageMode {
  DEFAULT,
  APPROVAL,
}

interface IProps {
  filterable?: boolean;
  title: string;
  mode?: PageMode;
}

const ProjectListView: React.FC<IProps> = ({ filterable, title, mode }) => {
  const lookupCodes = useCodeLookups();
  const agencies = useMemo(() => lookupCodes.getByType(API.AGENCY_CODE_SET_NAME), [lookupCodes]);
  const projectStatuses = useSelector<RootState, IStatus[]>(state => state.statuses as any);
  const keycloak = useKeycloakWrapper();
  const [deleteId, setDeleteId] = React.useState<string | undefined>();
  const agencyIds = useMemo(() => agencies.map(x => parseInt(x.id, 10)), [agencies]);
  const agencyOptions = (agencies ?? []).map(c => mapLookupCodeWithParentString(c, agencies));
  const statuses = (projectStatuses ?? []).map(c => mapStatuses(c));
  const columns = useMemo(() => cols, []);

  // We'll start our table without any data
  const [data, setData] = useState<IProject[] | undefined>(undefined);

  // Filtering and pagination state
  const [filter, setFilter] = useState<IProjectFilterState>({});
  const [clearSelected, setClearSelected] = useState(false);
  const [pageSize, setPageSize] = useState(10);
  const [pageIndex, setPageIndex] = useState(0);
  const [pageCount, setPageCount] = useState(0);
  const history = useHistory();

  // const [loading, setLoading] = useState(false);
  const fetchIdRef = useRef(0);

  // This will get called when the table needs new data
  const handleRequestData = useCallback(
    async ({ pageIndex, pageSize }: { pageIndex: number; pageSize: number }) => {
      setPageSize(pageSize);
      setPageIndex(pageIndex);
    },
    [setPageSize, setPageIndex],
  );

  // the following claims are passed through to the columns in order to determine whether to display delete option
  const isAdmin = keycloak.hasClaim(Claims.ADMIN_PROJECTS);
  const projectEditClaim = keycloak.hasClaim(Claims.PROJECT_EDIT);
  const user = `${keycloak.lastName}, ${keycloak.firstName}`;

  // Update internal state whenever the filter bar state changes
  const handleFilterChange = useCallback(
    (value: IProjectFilterState) => {
      (value as any).agencies
        ? setFilter({ ...value, agencies: (value as any)?.agencies })
        : setFilter({ ...value });
      if ((value as any).statusId) {
        setFilter({ ...value, statusId: (value as any).statusId?.map((x: any) => x) });
      } else {
        setFilter({ ...value });
        setClearSelected(!clearSelected);
      }
      setPageIndex(0); // Go to first page of results when filter changes
    },
    [setFilter, setPageIndex, clearSelected],
  );
  const onPageSizeChanged = useCallback(size => {
    setPageSize(size);
  }, []);

  const fetchData = useCallback(
    async ({
      pageIndex,
      pageSize,
      filter,
      agencyIds,
    }: {
      pageIndex: number;
      pageSize: number;
      filter: IProjectFilterState;
      agencyIds: number[];
    }) => {
      // Give this fetch an ID
      const fetchId = ++fetchIdRef.current;

      // Only update the data if this is the latest fetch
      if (fetchId === fetchIdRef.current && agencyIds?.length > 0) {
        setData(undefined);
        const query = getServerQuery({
          pageIndex,
          pageSize,
          filter: mode === PageMode.APPROVAL ? { ...filter, assessWorkflow: true } : filter,
        });
        const data = await service.getProjectList(query);

        // The server could send back total page count.
        // For now we'll just calculate it.
        setData(
          data.items.map((project: IProject) => {
            return project;
          }),
        );
        setPageCount(Math.ceil(data.total / pageSize));
      }
    },
    [setData, setPageCount, mode],
  );
  const dispatch = useDispatch();
  const route = history.location.pathname;

  // Listen for changes in pagination and use the state to fetch our new data
  useDeepCompareEffect(() => {
    route === '/projects/list' && dispatch(fetchProjectStatuses());
    fetchData({ pageIndex, pageSize, filter, agencyIds });
  }, [fetchData, pageIndex, pageSize, filter, agencyIds, dispatch, route]);

  const fetch = (accept: 'csv' | 'excel', reportType: 'generic' | 'spl') => {
    const query = getServerQuery({ pageIndex, pageSize, filter, agencyIds });
    return dispatch(
      download({
        url:
          reportType === 'generic'
            ? getProjectReportUrl({ ...query, all: true })
            : getProjectFinancialReportUrl({ ...query, all: true }),
        fileName: `${reportType === 'spl' ? 'pims-spl-report' : 'pims-projects'}.${
          accept === 'csv' ? 'csv' : 'xlsx'
        }`,
        actionType: 'projects-report',
        headers: {
          Accept: accept === 'csv' ? 'text/csv' : 'application/vnd.ms-excel',
        },
      }),
    );
  };

  const handleDelete = async () => {
    const project = data?.find(p => p.projectNumber === deleteId);
    if (project) {
      project.status = projectStatuses.find((x: any) => x.name === project.status)!;
      await service.deleteProject(project);
      setData(data?.filter(p => p.projectNumber !== project.projectNumber));
      setDeleteId(undefined);
    }
  };

  const initiateDelete = (projectNumber: string) => {
    setDeleteId(projectNumber);
  };

  const onRowClick = (row: IProject) => {
    const ReviewWorkflowStatuses = Object.keys(ReviewWorkflowStatus).map(
      (k: string) => (ReviewWorkflowStatus as any)[k],
    );
    if (ReviewWorkflowStatuses.includes(row.statusCode)) {
      if (keycloak.hasClaim(Claims.ADMIN_PROJECTS)) {
        history.push(`${row.statusRoute}?projectNumber=${row.projectNumber}`);
      } else {
        history.push(`/projects/summary?projectNumber=${row.projectNumber}`);
      }
    } else {
      history.push(`/dispose${row.statusRoute}?projectNumber=${row.projectNumber}`);
    }
  };

  const lazyLoadProperties = async (expandedRows: IProject[]) => {
    if (expandedRows.length > 0) {
      expandedRows = expandedRows.filter(x => x.properties.length === 0);
      const properties = await Promise.all(
        expandedRows.map(async project => await service.loadProperties(project.projectNumber)),
      );
      const projectPropertiesMap = properties.reduce((map: any, current: any) => {
        const ids = Object.keys(current);
        const projectId = ids[0];
        return { ...map, [projectId]: current[projectId] };
      }, {});
      setData(
        data?.map(d => {
          return !!projectPropertiesMap[d.projectNumber]
            ? { ...d, properties: projectPropertiesMap[d.projectNumber] }
            : d;
        }),
      );
    }
  };

  const fiscalYears = React.useMemo(() => {
    const startYear = new Date().getFullYear() - 10;
    return Array.from(Array(12).keys())
      .map(i => {
        var year = startYear + i;
        return { label: `${year - 1} / ${year}`, value: year };
      })
      .reverse();
  }, []);

  return (
    <Container fluid className="ProjectListView">
      <div className="filter-container">
        {filterable && (
          <FilterBar<IProjectFilterState>
            initialValues={initialValues}
            onChange={handleFilterChange}
          >
            <Col xs={2} className="bar-item">
              <Input field="name" placeholder="Search by project name or number" />
            </Col>
            <Col xs={2} className="bar-item">
              <ParentSelect
                field={'statusId'}
                options={statuses}
                clearSelected={clearSelected}
                setClearSelected={setClearSelected}
                enableMultiple
                filterBy={['label', 'parent']}
                placeholder="Enter a Status"
              />
            </Col>
            <Col xs={2} className="bar-item">
              <ParentSelect
                field="agencies"
                options={agencyOptions}
                filterBy={['code', 'label', 'parent']}
                placeholder="Enter an Agency"
              />
            </Col>
            <Col xs={1} className="bar-item">
              <Select field="fiscalYear" options={fiscalYears} placeholder="Fiscal Year" />
            </Col>
          </FilterBar>
        )}
      </div>
      <div className="ScrollContainer">
        {!!deleteId && (
          <GenericModal
            display={!!deleteId}
            cancelButtonText="Cancel"
            okButtonText="Yes, Delete"
            handleOk={handleDelete}
            handleCancel={() => setDeleteId(undefined)}
            title="Confirm Delete"
            message="Are you sure that you want to delete this project?"
          />
        )}
        <Container fluid className="TableToolbar">
          <h3 className="mr-4">{title}</h3>
          {keycloak.hasClaim(Claims.REPORTS_VIEW) && (
            <>
              <TooltipWrapper toolTipId="export-to-excel" toolTip="Export to Excel">
                <FileIcon>
                  <FaFileExcel
                    size={36}
                    data-testid="excel-icon"
                    onClick={() => fetch('excel', 'generic')}
                  />
                </FileIcon>
              </TooltipWrapper>
              <TooltipWrapper toolTipId="export-to-excel" toolTip="Export to CSV">
                <FileIcon>
                  <FaFileAlt
                    size={36}
                    data-testid="csv-icon"
                    onClick={() => fetch('csv', 'generic')}
                  />
                </FileIcon>
              </TooltipWrapper>
            </>
          )}
        </Container>
        <Table<IProject>
          name="projectsTable"
          clickableTooltip="View Disposal Project details"
          columns={
            mode === PageMode.APPROVAL
              ? columns()
              : columns(initiateDelete, isAdmin, projectEditClaim, user)
          }
          data={data || []}
          loading={data === undefined}
          onRequestData={handleRequestData}
          pageCount={pageCount}
          pageSize={pageSize}
          pageIndex={pageIndex}
          onRowClick={onRowClick}
          detailsPanel={{
            render: project => <Properties data={project.properties} />,
            icons: {
              open: <FaFolderOpen color="black" size={20} />,
              closed: <FaFolder color="black" size={20} />,
            },
            checkExpanded: (row, state) => !!state.find(x => x.projectNumber === row.projectNumber),
            onExpand: lazyLoadProperties,
            getRowId: row => row.projectNumber,
          }}
          onPageSizeChange={onPageSizeChanged}
        />
      </div>
    </Container>
  );
};

export const ProjectApprovalRequestListView = () => {
  return (
    <ProjectListView
      filterable={false}
      mode={PageMode.APPROVAL}
      title="Surplus Property Program Projects - Approval Requests"
    />
  );
};

export default () => <ProjectListView title="My Agency's Projects" filterable={true} />;
