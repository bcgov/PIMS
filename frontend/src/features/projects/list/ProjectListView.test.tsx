import ProjectListView from './ProjectListView';
import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { render, cleanup, act, wait } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { ILookupCode } from 'actions/lookupActions';
import * as API from 'constants/API';
import { Provider } from 'react-redux';
import * as reducerTypes from 'constants/reducerTypes';
import service from '../apiService';
import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import { useKeycloak } from '@react-keycloak/web';
import Claims from 'constants/claims';

const mockAxios = new MockAdapter(axios);
mockAxios.onAny().reply(200, {});

jest.mock('@react-keycloak/web');
const mockKeycloak = (claims: string[]) => {
  (useKeycloak as jest.Mock).mockReturnValue({
    keycloak: {
      userInfo: {
        agencies: [1],
        roles: claims,
      },
      subject: 'test',
    },
  });
};

const testData = {
  items: [
    {
      projectNumber: 'SPP-10015',
      name: 'Project name 16',
      statusId: 0,
      status: 'Draft',
      tierLevelId: 1,
      tierLevel: 'Tier 1',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
      note: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
      agencyId: 35,
      agency: 'AEST',
      subAgency: 'UOV',
      properties: [
        {
          id: 1,
          propertyTypeId: 1,
          description:
            'Gym, teaching kitchen, energy lab, greenhouse. Building full name: Center for Excellence in Sustainability',
          classification: 'Core Operational',
          agencyId: 0,
          subAgency: 'Nichola Valley Institute of Technology',
          agencyCode: 'AEST',
          address: '4155 Belshaw St',
          city: 'Merritt',
          netBook: 0,
          assessed: 0,
          market: 0,
          landArea: 26.9,
          createdOn: '2020-05-28T18:48:03.181977',
          rowVersion: 'AAAAAAAAfRA=',
        },
      ],
      createdBy: 'User, Administrator',
      createdOn: '2020-05-28T18:48:03.181976',
      rowVersion: 'AAAAAAAAfQ8=',
    },
  ],
  page: 1,
  quantity: 10,
  pageIndex: 0,
  total: 1,
};

// Set all module functions to jest.fn
jest.mock('../apiService');
const mockedService = service as jest.Mocked<typeof service>;

const mockStore = configureMockStore([thunk]);

const lCodes = {
  lookupCodes: [
    { name: 'agencyVal', id: '1', isDisabled: false, type: API.AGENCY_CODE_SET_NAME },
    {
      name: 'classificationVal',
      id: '1',
      isDisabled: false,
      type: API.PROPERTY_CLASSIFICATION_CODE_SET_NAME,
    },
  ] as ILookupCode[],
};

const store = mockStore({
  [reducerTypes.LOOKUP_CODE]: lCodes,
});

const history = createMemoryHistory();

describe('Project list view tests', () => {
  // clear mocks before each test
  beforeEach(() => {
    mockedService.getProjectList.mockClear();
    mockedService.deleteProject.mockClear();
    mockKeycloak([]);
  });
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('Matches snapshot', async () => {
    mockedService.getProjectList.mockResolvedValueOnce(testData as any);

    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <Router history={history}>
            <ProjectListView />
          </Router>
        </Provider>,
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  it('Displays message for empty list', async () => {
    mockedService.getProjectList.mockResolvedValueOnce({
      quantity: 0,
      total: 0,
      page: 1,
      pageIndex: 0,
      items: [],
    });

    await act(async () => {
      const { findByText, container } = render(
        <Provider store={store}>
          <Router history={history}>
            <ProjectListView />
          </Router>
        </Provider>,
      );

      // default table message when there is no data to display
      const noResults = await findByText('No rows to display');
      expect(noResults).toBeVisible();
      expect(container.querySelector('span[class="spinner-border"]')).not.toBeInTheDocument();
    });
  });

  it('Does not display export buttons by default', async () => {
    mockedService.getProjectList.mockResolvedValueOnce({
      quantity: 0,
      total: 0,
      page: 1,
      pageIndex: 0,
      items: [],
    });

    await act(async () => {
      const { queryByText, queryByTitle, container } = render(
        <Provider store={store}>
          <Router history={history}>
            <ProjectListView />
          </Router>
        </Provider>,
      );
      expect(queryByTitle('Export to Excel')).not.toBeInTheDocument();
      expect(queryByTitle('Export to CSV')).not.toBeInTheDocument();
      expect(queryByText('SPL Report')).not.toBeInTheDocument();
      expect(container.querySelector('span[class="spinner-border"]')).not.toBeInTheDocument();
    });
  });

  it('Displays export buttons with view reports permission', async () => {
    mockKeycloak([Claims.REPORTS_VIEW, Claims.PROJECT_VIEW]);
    mockedService.getProjectList.mockResolvedValueOnce({
      quantity: 0,
      total: 0,
      page: 1,
      pageIndex: 0,
      items: [],
    });

    await act(async () => {
      const { queryByTitle, container } = render(
        <Provider store={store}>
          <Router history={history}>
            <ProjectListView />
          </Router>
        </Provider>,
      );
      expect(queryByTitle('Export to Excel')).toBeInTheDocument();
      expect(queryByTitle('Export to CSV')).toBeInTheDocument();
      expect(container.querySelector('span[class="spinner-border"]')).not.toBeInTheDocument();
    });
  });

  it('Displays export buttons with spl reports permission', async () => {
    mockKeycloak([Claims.REPORTS_SPL]);
    mockedService.getProjectList.mockResolvedValueOnce({
      quantity: 0,
      total: 0,
      page: 1,
      pageIndex: 0,
      items: [],
    });

    await act(async () => {
      const { queryByText, container } = render(
        <Provider store={store}>
          <Router history={history}>
            <ProjectListView />
          </Router>
        </Provider>,
      );
      expect(queryByText('SPL Report')).toBeVisible();
      await wait(async () => {
        expect(container.querySelector('span[class="spinner-border"]')).not.toBeInTheDocument();
      });
    });
  });
});
