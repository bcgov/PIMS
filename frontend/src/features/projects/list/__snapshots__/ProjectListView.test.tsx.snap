// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project list view tests Matches snapshot 1`] = `
"<div class=\\"ProjectListView container-fluid\\">
  <div class=\\"filter-container\\">
    <form novalidate=\\"\\" class=\\"\\">
      <div class=\\"search-bar form-row\\">
        <h3 class=\\"filterBarHeading\\"></h3>
        <div class=\\"bar-item col-2\\">
          <div class=\\"form-group\\"><select name=\\"statusId\\" id=\\"input-statusId\\" class=\\"form-select form-control\\">
              <option value=\\"\\">Select a project status</option>
            </select></div>
        </div>
        <div class=\\"bar-item col-2\\">
          <div class=\\"form-group\\">
            <div class=\\"rbt\\" style=\\"outline: none; position: relative;\\" tabindex=\\"-1\\">
              <div style=\\"display: flex; flex: 1; height: 100%; position: relative;\\"><input autocomplete=\\"off\\" placeholder=\\"Enter an Agency\\" type=\\"text\\" name=\\"agencies\\" id=\\"agencies-field\\" aria-autocomplete=\\"both\\" aria-expanded=\\"false\\" aria-haspopup=\\"listbox\\" role=\\"combobox\\" class=\\"rbt-input-main form-control rbt-input map-filter-typeahead\\" value=\\"\\"><input aria-hidden=\\"true\\" class=\\"rbt-input-hint\\" readonly=\\"\\" style=\\"background-color: transparent; border-color: transparent; box-shadow: none; color: rgba(0, 0, 0, 0.35); left: 0px; pointer-events: none; position: absolute; top: 0px; width: 100%;\\" tabindex=\\"-1\\" value=\\"\\"></div>
            </div>
          </div>
        </div>
        <div class=\\"bar-item col-2\\">
          <div class=\\"form-group\\"><input name=\\"name\\" placeholder=\\"Search by project name or number\\" id=\\"input-name\\" class=\\"form-control\\" value=\\"\\"></div>
        </div>
        <div class=\\"bar-item flex-grow-0 col\\"><button type=\\"submit\\" class=\\"Button Button--icon-only bg-warning btn btn-primary\\">
            <div class=\\"Button__icon\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 512 512\\" size=\\"20\\" height=\\"20\\" width=\\"20\\" xmlns=\\"http://www.w3.org/2000/svg\\">
                <path d=\\"M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z\\"></path>
              </svg></div>
          </button></div>
        <div class=\\"bar-item flex-grow-0 col\\"><button type=\\"reset\\" class=\\"Button Button--icon-only btn btn-secondary\\">
            <div class=\\"Button__icon\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 512 512\\" size=\\"20\\" height=\\"20\\" width=\\"20\\" xmlns=\\"http://www.w3.org/2000/svg\\">
                <path d=\\"M212.333 224.333H12c-6.627 0-12-5.373-12-12V12C0 5.373 5.373 0 12 0h48c6.627 0 12 5.373 12 12v78.112C117.773 39.279 184.26 7.47 258.175 8.007c136.906.994 246.448 111.623 246.157 248.532C504.041 393.258 393.12 504 256.333 504c-64.089 0-122.496-24.313-166.51-64.215-5.099-4.622-5.334-12.554-.467-17.42l33.967-33.967c4.474-4.474 11.662-4.717 16.401-.525C170.76 415.336 211.58 432 256.333 432c97.268 0 176-78.716 176-176 0-97.267-78.716-176-176-176-58.496 0-110.28 28.476-142.274 72.333h98.274c6.627 0 12 5.373 12 12v48c0 6.627-5.373 12-12 12z\\"></path>
              </svg></div>
          </button></div>
      </div>
    </form>
  </div>
  <div class=\\"ScrollContainer\\">
    <div class=\\"TableToolbar container-fluid\\">
      <h3 class=\\"mr-4\\">My Agency's Projects</h3>
    </div>
    <div role=\\"table\\" class=\\"table\\">
      <div class=\\"thead thead-light\\">
        <div role=\\"row\\" style=\\"display: flex; flex: 1 0 auto; min-width: 875px;\\" class=\\"tr\\">
          <div class=\\"th expander svg-btn\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 576 512\\" color=\\"black\\" size=\\"20\\" style=\\"color: black;\\" height=\\"20\\" width=\\"20\\" xmlns=\\"http://www.w3.org/2000/svg\\">
              <path d=\\"M572.694 292.093L500.27 416.248A63.997 63.997 0 0 1 444.989 448H45.025c-18.523 0-30.064-20.093-20.731-36.093l72.424-124.155A64 64 0 0 1 152 256h399.964c18.523 0 30.064 20.093 20.73 36.093zM152 224h328v-48c0-26.51-21.49-48-48-48H272l-64-64H48C21.49 64 0 85.49 0 112v278.046l69.077-118.418C86.214 242.25 117.989 224 152 224z\\"></path>
            </svg></div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 65 0 auto; min-width: 65px; width: 3.5%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"3.5%\\" class=\\"th\\">Project No.</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 200 0 auto; min-width: 200px; width: 7%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"7%\\" class=\\"th\\">Name</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 50 0 auto; min-width: 50px; width: 7%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"7%\\" class=\\"th\\">Status</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 80 0 auto; min-width: 80px; width: 7%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"7%\\" class=\\"th\\">Agency</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 80 0 auto; min-width: 80px; width: 3.5%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"3.5%\\" class=\\"th\\">Zoning</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 80 0 auto; min-width: 80px; width: 3.5%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"3.5%\\" class=\\"th\\">Zoning Potential</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 80 0 auto; min-width: 80px; width: 3.5%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"3.5%\\" class=\\"th\\">Net Book</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 80 0 auto; min-width: 80px; width: 3.5%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"3.5%\\" class=\\"th\\">Market Value</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 80 0 auto; min-width: 80px; width: 7%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"7%\\" class=\\"th\\">Updated On</div>
          <div colspan=\\"1\\" role=\\"columnheader\\" style=\\"box-sizing: border-box; flex: 80 0 auto; min-width: 80px; width: 7%; justify-content: flex-start; text-align: left; align-items: center; display: flex;\\" width=\\"7%\\" class=\\"th\\">Updated By</div>
        </div>
      </div>
      <div class=\\"table-loading\\">
        <div class=\\"spinner-border\\"></div>
      </div>
    </div>
    <div class=\\"table-toolbar\\"></div>
  </div>
</div>"
`;
